### ISSUE TITLE
Raw endpoints

### ISSUE THREAD

**Issue by SzymonPobiega** [Particular/NServiceBus#3389]
Support for raw receive and send messages like spiked here: https://github.com/Particular/NServiceBus/pull/3206

A raw endpoint would have a super-simple receive pipeline with just the transport receive stage and a super-simple send pipeline with just the dispatch stage. It would be comparable to an externalized satellite. Usage scenarios:
- integration
- gateway replacement
- bridges
- proxies

**Comment by SzymonPobiega** [comment]
As @johnsimons pointed out, such an endpoint would be useful in ServiceControl where we need to receive raw messages from various queues. An advantage of raw endpoints over satellites would be ability to host them in separate processes.

//cc @danielmarbach

**Comment by timbussmann** [comment]
I generally like the idea of a very stripped down version of NSB, seems to fit well for the use cases you listed ( @indualagarsamy may confirm :D ). But instead of having a "regular" and a "micro" endpoint, wouldn't it make sense for the "regular" to build on top of the "raw" endpoint?

**Comment by SzymonPobiega** [comment]
@timbussmann that would be ideal but I think we are far from it because of the way our features are structured e.g. the pipelines are created in a monolithic fashion.

I had a dream some time ago that we build the pipelines by extending them. In such design we would start with *raw* endpoint that contains the incoming pipeline consisting of `TransportReceiveContext` and outgoing pipeline consisting of `DispatchPipeline`. Each feature in the *core* can extend one or both these pipelines by adding a stage. This way we would have feature like
 * Physical message processing
 * Logical message processing 
 * Routing
that extend the pipeline and allow other features to be hooked into proper places.

**Comment by timbussmann** [comment]
I know @indualagarsamy is working on a bridge implementation and it's quite a different implementation from v5. Would be interesting to compare the "regular endpoint" bridge implementation and the "micro endpoint bridge implementation".

**Comment by indualagarsamy** [comment]
@timbussmann @SzymonPobiega - What I am doing for the SqlBridge after evaluating a couple of approaches is just plain old MSMQ and plain old Sql. 

I have a working sample, but (still a WIP). But essentially what I have is a very simple process (has nothing to do with NServiceBus). Uses MSMQ to receive messages using native messaging API. Gets the body as byte array, gets the headers and then uses Sql to send messages to the appropriate queue using the already documented: http://docs.particular.net/nservicebus/sqlserver/operations-scripting#native-send-the-native-send-helper-methods. 

No micro-endpoints, no stripped down endpoints. Just bare metal transport code. 
Thanks @SimonCropp for the idea.

**Comment by andreasohlund** [comment]
>A raw endpoint would have a super-simple receive pipeline with just the transport receive stage

Is the v6 satellites + access to `IDispatchMessages` close enough? 

Is there anything actionable here or should we close this for now?

**Comment by danielmarbach** [comment]
I think we would need a better more explicit way to access a dispatcher than `builder.Build`

**Comment by andreasohlund** [comment]
>I think we would need a better more explicit way to access a dispatcher than builder.Build

Should we raise a new issue to deal with this?

**Comment by SzymonPobiega** [comment]
I'll raise a new issue.

**Comment by timbussmann** [comment]
@SzymonPobiega I wonder what parts of the core does your "raw endpoint" really need? Aren't you essentially thinking of a small, dedicated layer on top of the transport implementations?

**Comment by SzymonPobiega** [comment]
@timbussmann yes, indeed. I am still on the fence if a raw endpoint should use the concept of feature or not. If yes, then regular endpoint is just a raw endpoint with additional features. Otherwise a raw endpoint would require its own `IntializableEndpoint` and `StartableEndpoint` equivalents.

I think I'll prepare two variants and then we can compare the pros and cons of each.

**Comment by SzymonPobiega** [comment]
I've quickly spiked the "regular endpoint as an extension of a raw endpoint" approach to asses how much effort would it require

 * Make `EndpointConfiguration` generic of `TInt` and `TImpl` where `TInt` is the API of the endpoint (equivalent to `IMessageSession` for regular endpoints) and `TImpl` is the implementation of that API
 * Force derived `EndpointConfigurations` to provide a factory method for creating an instance of `TImpl` that startable endpoint can use
 * Force derived `EndpointConfigurations` to provide a list of features enabled by default in a given mode
 * Make `IStartableEndpoint` generic of `TInt`
 * Make `InitializableEndpoint` and `StartableEndpoint` generic
 * `RunningEndpointInstance` would have to implement `TInt` by mixing in `TImpl` instance
 * `FeatureStartupTask` would have to inherit from a generic `FST<TInt>`
 * `FeatureConfigurationContext` API would have to either be lax when it comes to FST typing or it also would have to be generic, forcing `Feature` to be generic too.

To sum up, it seems like a very interesting academic project to make everything generic based on the shape of the API of the endpoint but...

it would be *much easier* to just create a bunch of specific classes that implement raw endpoints. Another benefit of "decoupled" approach would be that such raw endpoints potentially might not need a container  and FST are overkill for raw endpoints anyway.

**Comment by SzymonPobiega** [comment]
@timbussmann I now *think* I can implement it outside of the core entirely.

**Comment by SzymonPobiega** [comment]
Here's a problem list when trying to implement a raw endpoint outside of the core:

```
error CS1061: 'CriticalError' does not contain a definition for 'SetEndpoint' and no extension method 'SetEndpoint' accepting a first argument of type 'CriticalError' could be found (are you missing a using directive or an assembly reference?)
error CS0122: 'TransportConnectionString' is inaccessible due to its protection level
error CS0122: 'TransportConnectionString.GetConnectionStringOrRaiseError(TransportDefinition)' is inaccessible due to its protection level
error CS1061: 'IncomingMessage' does not contain a definition for 'RevertToOriginalBodyIfNeeded' and no extension method 'RevertToOriginalBodyIfNeeded' accepting a first argument of type 'IncomingMessage' could be found (are you missing a using directive or an assembly reference?)
error CS0122: 'ExceptionHeaderHelper' is inaccessible due to its protection level
error CS1061: 'TransportSendInfrastructure' does not contain a definition for 'PreStartupCheck' and no extension method 'PreStartupCheck' accepting a first argument of type 'TransportSendInfrastructure' could be found (are you missing a using directive or an assembly reference?)
error CS1061: 'TransportReceiveInfrastructure' does not contain a definition for 'PreStartupCheck' and no extension method 'PreStartupCheck' accepting a first argument of type 'TransportReceiveInfrastructure' could be found (are you missing a using directive or an assembly reference?)
```

Neither of these is a critical and these access problems can be worked around via reflection (not on the hot path)

**Comment by SzymonPobiega** [comment]
I implemented it myself. Closing.
